tsdb/head.go

```diff
@@ -525,27 +525,57 @@ Outer:
 					break Outer
 				}
 
-				if created {
-					// If this series gets a duplicate record, we don't restore its mmapped chunks,
-					// and instead restore everything from WAL records.
-					series.mmappedChunks = mmappedChunks[series.ref]
+				if h.lastSeriesID.Load() < s.Ref {
+					h.lastSeriesID.Store(s.Ref)
+				}
 
-					h.metrics.chunks.Add(float64(len(series.mmappedChunks)))
-					h.metrics.chunksCreated.Add(float64(len(series.mmappedChunks)))
+				if !created {
+					// There's already a different ref for this series.
+					multiRef[s.Ref] = series.ref
+				}
+
+				// A new series record is only possible when the old samples were already compacted into a block.
+				// Hence we can discard all the samples and m-mapped chunks replayed till now for this series.
+				mmc := mmappedChunks[s.Ref]
+				if len(mmc) == 0 {
+					// We continue with the old data if there is nothing to overwrite.
+					// Stale data will be removed during garbage collection.
+					continue
+				}
 
-					if len(series.mmappedChunks) > 0 {
-						h.updateMinMaxTime(series.minTime(), series.maxTime())
+				if len(series.mmappedChunks) > 0 {
+					if overlapsClosedInterval(
+						series.mmappedChunks[0].minTime,
+						series.mmappedChunks[len(series.mmappedChunks)-1].maxTime,
+						mmc[0].minTime,
+						mmc[len(mmc)-1].maxTime,
+					) {
+						// The m-map chunks for the new series ref overlaps with old m-map chunks.
```

Comment:
This should never happen, right? Worth commenting this?

Reply:
In an ideal world this would never happen. But if someone is tinkering around the data dir, this is possible. I will comment that this should not happen unless you are experimenting with something.